airbyte-workers/src/main/java/io/airbyte/workers/process/IntegrationLauncher.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.workers.process;
+
+import io.airbyte.workers.WorkerException;
+import java.nio.file.Path;
+
+public interface IntegrationLauncher {
+
+  ProcessBuilder spec(final Path jobRoot) throws WorkerException;
+
+  ProcessBuilder check(final Path jobRoot, final String configFilename) throws WorkerException;
+
+  ProcessBuilder discover(final Path jobRoot, final String configFilename) throws WorkerException;
+
+  ProcessBuilder read(final Path jobRoot,
+                      final String configFilename,
+                      final String catalogFilename,
+                      final String stateFilename)
```

Comment:
should this be optional instead of passing it as null?

Reply:
Not a fan of passing optionals as parameters, overloading of methods work very well for that.

There is no real rule but it seems generally accepted that Optionals are better used for return types to indicate the absence of a response (vs a null response or the actual response).

See Brian Goetz take on it https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555

Reply:
interesting take -- thanks for sharing